### PR TITLE
feat(api): remove dal dependencies from workflow queue service

### PR DESCRIPTION
**Pull request by p-fernandez**
### What change does this PR introduce?

Extracts some logic from the Workflow Queue service that led to DAL dependency and inverts the control through dependency injection of said logic inside use cases.
Also adds some tests for the job status update callbacks.

### Why was this change needed?

Proper inversion control and move forward to make the Workflow Queue service domain logic agnostic, so in the future when replaced by Kakfka/Pulsar (@Cliftonz is investigating), the replacement should be easier to do.

### CODE REVIEW

File: apps/api/src/app/events/services/workflow-queue/workflow.queue.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+  it('should a job added to the queue be updated as failed if it fails', async () => {
+    const transactionId = uuid();
+    const _environmentId = session.environment._id;
+    const _notificationId = NotificationRepository.createObjectId();
+    const _organizationId = session.organization._id;
+    const _subscriberId = subscriber._id;
+    const _templateId = NotificationTemplateRepository.createObjectId();
+    const _userId = session.user._id;
+
+    // Job with fake notification template and fake notification.
```
Comment (p-fernandez) [inline apps/api/src/app/events/services/workflow-queue/workflow.queue.service.spec.ts:132]: Job created on purpose to trigger a failure and the workflow queue service `failed` event listener callback.

File: apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts [hunk 2]
```diff
@@ -32,40 +40,45 @@ export class WorkflowQueueService {
...
   @Inject()
   private runJob: RunJob;
+  @Inject()
+  private setJobAsCompleted: SetJobAsCompleted;
+  @Inject()
+  private setJobAsFailed: SetJobAsFailed;
   private readonly queueScheduler: QueueScheduler;
   readonly DEFAULT_ATTEMPTS = 3;
 
-  constructor(private createExecutionDetails: CreateExecutionDetails, private queueNextJob: QueueNextJob) {
```
Comment (p-fernandez) [inline apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts:41]: Let's take advantage of Nest.js dependency injection functionality.

File: apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts [hunk 3]
```diff
@@ -92,10 +105,47 @@ export class WorkflowQueueService {
...
+          organizationId: job.data._organizationId,
+        })
+      );
+    }
+
+    const lastWebhookFilterRetry = job.attemptsMade === this.DEFAULT_ATTEMPTS && hasToBackoff;
+
+    if (lastWebhookFilterRetry) {
+      await this.handleLastFailedWebhookFilter(job, error);
+    }
```
Comment (p-fernandez) [inline apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts:136]: Once understanding what's happening here and found the proper naming, this should be moved to an use case to finalise the domain logic extraction.

File: apps/api/src/app/events/usecases/run-job/run-job.usecase.ts [hunk 4]
```diff
@@ -2,14 +2,15 @@ import { Injectable } from '@nestjs/common';
...
+import { SendMessage, SendMessageCommand } from '../send-message';
+
 import { ApiException } from '../../../shared/exceptions/api.exception';
 import { StorageHelperService } from '../../services/storage-helper-service/storage-helper.service';
-import { QueueNextJobCommand } from '../queue-next-job/queue-next-job.command';
-import { QueueNextJob } from '../queue-next-job/queue-next-job.usecase';
-import { SendMessageCommand } from '../send-message/send-message.command';
-import { SendMessage } from '../send-message/send-message.usecase';
-import { RunJobCommand } from './run-job.command';
-import { shouldBackoff } from '../../services/workflow-queue/workflow.queue.service';
```
Comment (p-fernandez) [inline apps/api/src/app/events/usecases/run-job/run-job.usecase.ts:12]: Probably thanks to Nest.js dependency injection this circular dependency didn't trigger off.

File: apps/api/src/app/events/usecases/update-job-status/update-job-status.usecase.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import { JobRepository } from '@novu/dal';
+
+import { UpdateJobStatusCommand } from './update-job-status.command';
+
+@Injectable()
+export class UpdateJobStatus {
+  constructor(private jobRepository: JobRepository) {}
+
+  public async execute(command: UpdateJobStatusCommand): Promise<void> {
+    await this.jobRepository.updateStatus(command.organizationId, command._jobId, command.status);
```
Comment (p-fernandez) [inline apps/api/src/app/events/usecases/update-job-status/update-job-status.usecase.ts:11]: Someone would say it is ridiculous to have this small use case... I'd rather have this simple and functionalities separated (and above all, out of the workflow queue service) and be able to easily modify them as needed.

File: apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts (hunk 3 above)
Comment (tsssdev) [reply apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts:136]: All he was trying to do is, not to fail the job when number of attempts  not reached(number of attempts 3 for jobs with webhook filters), so those jobs should skip the failure path for first and 2nd failure and on 3rd failure(last failure) it should go to regular failure path and decide to schedule next job or no, based on the flag shouldStopOnFailt.. this is also taken care very cleanly in #2727 and much simpler to understand in new code

File: apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts [hunk 6]
```diff
@@ -1,18 +1,26 @@
...
 import {
   CreateExecutionDetailsCommand,
   DetailEnum,
 } from '../../../execution-details/usecases/create-execution-details/create-execution-details.command';
-import { EXCEPTION_MESSAGE_ON_WEBHOOK_FILTER } from '../../../shared/constants';
-import { QueueNextJobCommand } from '../../usecases/queue-next-job/queue-next-job.command';
-import { QueueNextJob } from '../../usecases/queue-next-job/queue-next-job.usecase';
-import { ConnectionOptions } from 'tls';
+
+export const WORKER_NAME = 'standard';
```
Comment (scopsy) [inline apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts:23]: Great addition

File: apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts [hunk 7]
```diff
@@ -32,40 +40,45 @@ export class WorkflowQueueService {
...
-  private jobRepository: JobRepository;
+  private createExecutionDetails: CreateExecutionDetails;
+  @Inject()
+  private queueNextJob: QueueNextJob;
   @Inject()
   private runJob: RunJob;
+  @Inject()
+  private setJobAsCompleted: SetJobAsCompleted;
+  @Inject()
+  private setJobAsFailed: SetJobAsFailed;
```
Comment (scopsy) [inline apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts:51]: Would suggest moving all of those to the constructor with @Inject, easier to read in my opinion and follows the pattern of all our other services

File: apps/api/src/app/events/usecases/update-job-status/update-job-status.usecase.ts (hunk 5 above)
Comment (scopsy) [reply apps/api/src/app/events/usecases/update-job-status/update-job-status.usecase.ts:11]: I think that's totally fine here.

**Review by scopsy — APPROVED** [review]
Looks good! Added a few minor comments

File: apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts (hunk 7 above)
Comment (p-fernandez) [reply apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts]: Doing this way we are forced to use the `forwardRef` functionality as injecting them as arguments generates a circular dependency with the Events module due `AddJob` depends on the Workflow Queue service.

**Pull request merged**